platforms/software/dependency-management/src/main/java/org/gradle/internal/resolve/resolver/FilteringResolvedArtifactSet.java

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.resolve.resolver;
+
+import org.gradle.api.Action;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ArtifactVisitor;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvableArtifact;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ResolvedArtifactSet;
+import org.gradle.api.internal.artifacts.transform.TransformStepNode;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.FileCollectionStructureVisitor;
+import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
+import org.gradle.internal.DisplayName;
+import org.gradle.internal.component.external.model.ImmutableCapabilities;
+import org.gradle.internal.operations.BuildOperationQueue;
+import org.gradle.internal.operations.RunnableBuildOperation;
+
+import java.util.function.Predicate;
+
+/**
+ * A {@link ResolvedArtifactSet} that applies a filter to the artifacts of a delegate {@link ResolvedArtifactSet}.
+ * <p>
+ * The filter is applied <strong>after</strong> build dependencies are calculated, meaning the filter is not
+ * applied to build dependencies. This is because the filter may be a function of the resolved artifact files,
+ * which are not known until after build dependencies are executed.
+ */
+class FilteringResolvedArtifactSet implements ResolvedArtifactSet {
+
+    private final ResolvedArtifactSet artifacts;
+    private final Predicate<ResolvableArtifact> filter;
+
+    public FilteringResolvedArtifactSet(ResolvedArtifactSet artifacts, Predicate<ResolvableArtifact> filter) {
+        this.artifacts = artifacts;
+        this.filter = filter;
+    }
+
+    @Override
+    public void visit(Visitor visitor) {
+        artifacts.visit(new Visitor() {
+            @Override
+            public FileCollectionStructureVisitor.VisitType prepareForVisit(FileCollectionInternal.Source source) {
+                return visitor.prepareForVisit(source);
+            }
+
+            @Override
+            public void visitArtifacts(Artifacts artifacts) {
+                visitor.visitArtifacts(new Artifacts() {
+                    @Override
+                    public void prepareForVisitingIfNotAlready() {
+                        artifacts.prepareForVisitingIfNotAlready();
+                    }
+
+                    @Override
+                    public void startFinalization(BuildOperationQueue<RunnableBuildOperation> actions, boolean requireFiles) {
+                        artifacts.startFinalization(actions, requireFiles);
+                    }
+
+                    @Override
+                    public void visit(ArtifactVisitor visitor) {
+                        artifacts.visit(new ArtifactVisitor() {
```

Comment:
❗ After hitting the 3rd level of nested anonymous implementation, I really think these ought to be named types.

This would have the benefits of allowing them to be `static` types, and make the logic clearer when following the logic when debugging/looking at stack traces.

Reply:
yeah this was hard to write and read for myself as well. We should really make a `ResolvedArtifactSet` easier to work with so you don't need to many visitor types.

Changed.